comb_spec_searcher/class_queue.py

```diff
@@ -165,37 +154,38 @@ def _populate_staging(self) -> None:
         while not self.staging and self.working:
             self.staging.extend(self._iter_helper_working())
         while not self.staging:
-            if not self.curr_level:
+            if not any(self.curr_level):
                 self._change_level()
             self.staging.extend(self._iter_helper_curr())
 
     def _change_level(self) -> None:
         assert not self.staging, "Can't change level is staging is not empty"
         assert not self.working, "Can't change level is working is not empty"
-        assert not self.curr_level, "Can't change level is curr_level is not empty"
-        self.curr_level = deque(
-            label
-            for label, _ in sorted(self.next_level.items(), key=lambda x: -x[1])
-            if self.expansion_strats and self.can_do_expansion(label, 0)
+        assert not any(self.curr_level), "Can't change level is curr_level is not empty"
+        self.curr_level[0].extend(
+            label for label, _ in sorted(self.next_level.items(), key=lambda x: -x[1])
         )
-        if not self.curr_level:
+        if not any(self.curr_level):
             raise StopIteration
-        self.queue_sizes.append(len(self.curr_level))
+        self.queue_sizes.append(len(self.curr_level[0]))
         self.next_level = Counter()
 
     def _iter_helper_curr(self) -> Iterator[WorkPacket]:
-        label = self.curr_level.popleft()
-        for idx, strats in enumerate(self.expansion_strats):
-            if self.can_do_expansion(label, idx):
-                for strat in strats:
-                    yield WorkPacket(label, (strat,), False)
-                self.set_expansion_expanded(label, idx)
-                if self.can_do_expansion(label, idx + 1):
-                    self.curr_level.append(label)
-                break
-        else:
-            # finished applying all strategies to this label, ignore from now on
+        assert any(self.curr_level), "The current queue is empty"
+        # pylint: disable=stop-iteration-return
+        idx, label = next(
+            (
+                (idx, queue.popleft())
+                for idx, queue in enumerate(self.curr_level)
+                if queue
+            )
+        )
+        if idx == len(self.expansion_strats):
             self.set_stop_yielding(label)
+            return
+        for strat in self.expansion_strats[idx]:
+            yield WorkPacket(label, (strat,), False)
+        self.curr_level[idx + 1].append(label)
```

Comment:
You have `if idx == len(self.expansion_strats` then stop yielding, but instead, you could just not add it to the "extra" queue at the end.  (so instead if `idx < len(self.expansion_strats) - 1` add to the queue, else stop yielding.

Reply:
After discussing privately, the approach used is due to the staging, so we would be setting to ignore too soon.